In [8]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from scipy.stats import uniform, norm, expon

Branch banks must keep enough money on hand to satisfy customers' cash demands.  An armored truck delivers cash to the bank once a week.  The bank manager can choose the amount of weekly cash to have delivered.  Running out of cash during the week is terrible customer service and the manager wants to avoid this.  On the other hand, keeping excessive cash reserves costs the bank profits, since cash is a non-interest earning asset.  

The daily demand for cash at this particular bank follows a normal distribution with daily means and std dev summarized in Table 1.

**Table 1**

|                   |Monday|Tuesday|Wednesday|Thursday|Friday|Saturday|Sunday|
|:-----------------:|:----:|:-----:|:-------:|:------:|:----:|:------:|:----:|
|Mean (\\$1,000s)   |175   |120    |90       |60      |120   |140     |65    |
|Std Dev (\\$1,000s)|26    |18     |13       |9       |18    |21      |9     |  

  
  
  
a.  Suppose the bank manager starts the week with \\$825,000.  Create a model for daily ending balance and use 10,000 simluations to determine the probability of the bank running out of money at some point during the week.


b.  Estimate the amount of money needed at the start of the week to ensure that there is at most a 2.0\% probability of running out of money.


In [9]:
means = [175, 120, 90, 60, 120, 140, 65]
sds = [26, 18, 13, 9, 18, 21, 9]

sds

[26, 18, 13, 9, 18, 21, 9]

In [10]:
means[0]

175

In [11]:
## Simulation model for Cash Balance
def simulate_daily_ending_balance(initial_balance, daily_means, daily_std_devs, n_simulations):
    # Initialize the count of simulations where the bank runs out of money
    shortages = 0
    # Perform the simulations
    for _ in range(n_simulations):
        balance = initial_balance
        # Simulate the cash demands for each day of the week
        for mean, std_dev in zip(daily_means, daily_std_devs):
            daily_demand = np.random.normal(mean * 1000, std_dev * 1000)
            balance -= daily_demand
            # Check if the bank runs out of money
            if balance < 0:
                shortages += 1
                break
    # Calculate the probability of running out of money
    probability_of_shortage = shortages / n_simulations
    return probability_of_shortage


In [12]:
## Simulation Trials

import numpy as np

def simulation_trials(initial_balance, daily_means, daily_std_devs, n_simulations):
    # This function will store each trial's result
    trials_results = []
    for _ in range(n_simulations):
        result = simulate_daily_ending_balance(initial_balance, daily_means, daily_std_devs, 1)
        trials_results.append(result)
    # Return the list of results
    return trials_results

# Parameters
initial_balance = 825000  # Starting balance for the week
daily_means = np.array([175, 120, 90, 60, 120, 140, 65])  # Daily means in $1,000s
daily_std_devs = np.array([26, 18, 13, 9, 18, 21, 9])  # Daily standard deviations in $1,000s
n_simulations = 10000  # Number of simulations

# Execute the simulation model for cash balance
probability_of_shortage = simulate_daily_ending_balance(initial_balance, daily_means, daily_std_devs, n_simulations)

# Execute the simulation trials to collect individual trial results
trials_results = simulation_trials(initial_balance, daily_means, daily_std_devs, n_simulations)

# Output the probability of shortage from the simulation model
print(f"Probability of the bank running out of money: {probability_of_shortage:.4f}")


Probability of the bank running out of money: 0.1154


a.  Suppose the bank manager starts the week with \\$825,000.  Create a model for daily ending balance and use 10,000 simluations to determine the probability of the bank running out of money at some point during the week.

Answer Here_____

In [13]:
## Balance What-If
import numpy as np

# Function to simulate daily ending balance
def simulate_daily_ending_balance(initial_balance, daily_means, daily_std_devs, n_simulations):
    shortages = 0
    for _ in range(n_simulations):
        balance = initial_balance
        for mean, std_dev in zip(daily_means, daily_std_devs):
            daily_demand = np.random.normal(mean * 1000, std_dev * 1000)
            balance -= daily_demand
            if balance < 0:
                shortages += 1
                break
    probability_of_shortage = shortages / n_simulations
    return probability_of_shortage

# Function to find the minimum balance required to ensure at most a 2% probability of running out of money
def find_minimum_balance(daily_means, daily_std_devs, n_simulations, target_probability):
    low_balance = 0
    high_balance = 1000000  # Assuming a high initial balance that is unlikely to be reached
    tolerance = 1000  # Tolerance for binary search stopping criterion
    
    while low_balance + tolerance < high_balance:
        mid_balance = (low_balance + high_balance) / 2
        probability = simulate_daily_ending_balance(mid_balance, daily_means, daily_std_devs, n_simulations)
        if probability > target_probability:
            low_balance = mid_balance
        else:
            high_balance = mid_balance
            
    return high_balance

# Daily means and standard deviations for cash demand
daily_means = np.array([175, 120, 90, 60, 120, 140, 65])
daily_std_devs = np.array([26, 18, 13, 9, 18, 21, 9])
n_simulations = 10000
target_probability = 0.02  # 2% target probability

# Estimate the minimum balance required
minimum_balance_required = find_minimum_balance(daily_means, daily_std_devs, n_simulations, target_probability)
print(f"Minimum starting balance required: ${minimum_balance_required:,.2f}")



Minimum starting balance required: $863,281.25


b. Estimate the amount of money needed at the start of the week to ensure that there is at most a 2.0% probability of running out of money.

Answer Here $863,281.25